File: iOS/LiveStreamCore/Sources/Manager/Module/BattleManager.swift
//
//  BattleManager.swift
//  LiveStreamCore
//
//  Created by adamsfliu on 2024/12/19.
//

import RTCCommon
import RTCRoomEngine

class BattleManager {
    let observerState = ObservableState<BattleState>(initialState: BattleState())
    var battleState: BattleState {
        observerState.state
    }
    
    private weak var context: LiveStreamManager.Context?
    private let service: LiveStreamService
    
    init(context: LiveStreamManager.Context) {
        self.context = context
        self.service = context.service
    }
    
    func requestBattle(config: TUIBattleConfig, userIdList: [String],
                       timeout: TimeInterval) async throws -> (String, [TUIBattleUser]) {
        do {
            let result = try await service.requestBattle(config: config, userIdList: userIdList, timeout: timeout)
            modifyBattleState(value: result.0.battleId, keyPath: \BattleState.battleId)
            
            let userIds = result.1.filter {
                $0.value.intValue == TUIBattleCode.success.rawValue
            }.map { $0.key }
            
            context?.coHostManager.coHostState.connectedUserList.filter { user in
                userIds.contains(user.userId)
            }.map { user in
                let battleUser = TUIBattleUser()
                battleUser.roomId = user.roomId
                battleUser.userId = user.userId
                battleUser.userName = user.userName
                battleUser.avatarUrl = user.avatarUrl
                return battleUser
            }.forEach({ battleUser in
                addSentBattleRequest(user: battleUser)
            })
            return (result.0.battleId, battleState.sentBattleRequestList)
        } catch let LiveStreamCoreError.error(code, message) {
            LiveStreamLog.error("\(#file)","\(#line)","requestBattle:[code:\(code),message:\(message)]")
            throw LiveStreamCoreError.error(code: code, message: message)
        } catch {
            assert(false, "unknown error, description:[:\(error.localizedDescription)]")
            LiveStreamLog.error("\(#file)","\(#line)","unknown error, description:[:\(error.localizedDescription)]")
            return ("", [])
        }
    }
    
    func cancelBattle(battleId: String, userIdList: [String]) async throws {
        let battleUserIdList = battleState.sentBattleRequestList.filter { battleUser in
            userIdList.contains(battleUser.userId)
        }.map { $0.userId }
        
        do {
            try await service.cancelBattle(battleId: battleId, userIdList: battleUserIdList)
        } catch let LiveStreamCoreError.error(code, message) {
            LiveStreamLog.error("\(#file)","\(#line)","cancelBattle:[code:\(code),message:\(message)]")
            throw LiveStreamCoreError.error(code: code, message: message)
        } catch {
            assert(false, "unknown error, description:[:\(error.localizedDescription)]")
        }
    }
    
    func acceptBattle(battleId: String) async throws {
        do {
            try await service.acceptBattle(battleId: battleId)
        } catch let LiveStreamCoreError.error(code, message) {
            LiveStreamLog.error("\(#file)","\(#line)","acceptBattle:[code:\(code),message:\(message)]")
            throw LiveStreamCoreError.error(code: code, message: message)
        } catch {
            assert(false, "unknown error, description:[:\(error.localizedDescription)]")
        }
    }
    
    func rejectBattle(battleId: String) async throws {
        do {
            try await service.rejectBattle(battleId: battleId)
            modifyBattleState(value: [], keyPath: \BattleState.sentBattleRequestList)
        } catch let LiveStreamCoreError.error(code, message) {
            LiveStreamLog.error("\(#file)","\(#line)","rejectBattle:[code:\(code),message:\(message)]")
            throw LiveStreamCoreError.error(code: code, message: message)
        } catch {
            assert(false, "unknown error, description:[:\(error.localizedDescription)]")
        }
    }
    
    func exitBattle(battleId: String) async throws {
        do {
            try await service.exitBattle(battleId: battleId)
            modifyBattleState(value: [], keyPath: \BattleState.battledUsers, isPublished: true)
            modifyBattleState(value: "", keyPath: \BattleState.battleId)
        } catch let LiveStreamCoreError.error(code, message) {
            LiveStreamLog.error("\(#file)","\(#line)","rejectBattle:[code:\(code),message:\(message)]")
            throw LiveStreamCoreError.error(code: code, message: message)
        } catch {
            assert(false, "unknown error, description:[:\(error.localizedDescription)]")
        }
    }
}

// MARK: - Observer
extension BattleManager {
    func onBattleStarted(battleInfo: TUIBattleInfo) {
        modifyBattleState(value: battleInfo.battleId, keyPath: \BattleState.battleId)
        modifyBattleState(value: true, keyPath: \BattleState.isBattleStart, isPublished: true)
        var battleUserList = battleInfo.inviteeList
        battleUserList.append(battleInfo.inviter)
        modifyBattleState(value: battleUserList, keyPath: \BattleState.battledUsers, isPublished: true)
    }
    
    func onBattleEnded(battleInfo: TUIBattleInfo, reason: TUIBattleStoppedReason) {
        modifyBattleState(value: "", keyPath: \BattleState.battleId)
        modifyBattleState(value: false, keyPath: \BattleState.isBattleStart, isPublished: true)
        modifyBattleState(value: [], keyPath: \BattleState.battledUsers, isPublished: true)
    }
    
    func onUserJoinBattle(battleId: String, battleUser: TUIBattleUser) {
        guard battleId == battleState.battleId else { return }
        addBattleUser(user: battleUser)
    }
    
    func onUserExitBattle(battleId: String, battleUser: TUIBattleUser) {
        guard battleId == battleState.battleId else { return }
        removeBattleUser(user: battleUser)
    }
    
    func onBattleScoreChanged(battleId: String, battleUserList: [TUIBattleUser]) {
        
    }
    
    func onBattleRequestReceived(battleInfo: TUIBattleInfo, inviter: TUIBattleUser, invitee: TUIBattleUser) {
        modifyBattleState(value: battleInfo.battleId, keyPath: \BattleState.battleId)
    }
    
    func onBattleRequestCancelled(battleInfo: TUIBattleInfo, inviter: TUIBattleUser, invitee: TUIBattleUser) {
        modifyBattleState(value: "", keyPath: \BattleState.battleId)
    }
    
    func onBattleRequestTimeout(battleInfo: TUIBattleInfo, inviter: TUIBattleUser, invitee: TUIBattleUser) {
        modifyBattleState(value: "", keyPath: \BattleState.battleId)
    }
    
    func onBattleRequestAccept(battleInfo: TUIBattleInfo, inviter: TUIBattleUser, invitee: TUIBattleUser) {
        modifyBattleState(value: battleInfo.battleId, keyPath: \BattleState.battleId)
    }
    
    func onBattleRequestReject(battleInfo: TUIBattleInfo, inviter: TUIBattleUser, invitee: TUIBattleUser) {
        modifyBattleState(value: "", keyPath: \BattleState.battleId)
    }
}

// MARK: - Private
extension BattleManager {
    private func modifyBattleState<T>(value: T, keyPath: WritableKeyPath<BattleState, T>, isPublished: Bool = false) {
        observerState.update(isPublished: isPublished) { battleState in
            battleState[keyPath: keyPath] = value
        }
    }
    
    private func addBattleUser(user: TUIBattleUser) {
        observerState.update { battleState in
            battleState.battledUsers.append(user)
        }
    }
    
    private func removeBattleUser(user: TUIBattleUser) {
        observerState.update { battleState in
            if battleState.battledUsers.count == 2 { return }
            battleState.battledUsers.removeAll(where: {$0.userId == user.userId})
        }
    }
    
    private func addSentBattleRequest(user: TUIBattleUser) {
        observerState.update { battleState in
            battleState.sentBattleRequestList.append(user)
        }
    }
}

File: Android/livestreamcore/src/main/java/com/trtc/uikit/livekit/livestreamcore/manager/module/BattleManager.java
package com.trtc.uikit.livekit.livestreamcore.manager.module;

import android.content.Context;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.tencent.cloud.tuikit.engine.common.TUICommonDefine;
import com.tencent.cloud.tuikit.engine.extension.TUILiveBattleManager;
import com.tencent.cloud.tuikit.engine.extension.TUILiveBattleManager.BattleCode;
import com.tencent.cloud.tuikit.engine.extension.TUILiveBattleManager.BattleConfig;
import com.tencent.cloud.tuikit.engine.extension.TUILiveBattleManager.BattleInfo;
import com.tencent.cloud.tuikit.engine.extension.TUILiveBattleManager.BattleStoppedReason;
import com.tencent.cloud.tuikit.engine.extension.TUILiveBattleManager.BattleUser;
import com.tencent.cloud.tuikit.engine.extension.TUILiveConnectionManager.ConnectionUser;
import com.tencent.cloud.tuikit.engine.room.TUIRoomDefine;
import com.trtc.tuikit.common.system.ContextProvider;
import com.trtc.uikit.livekit.livestreamcore.LiveCoreViewDefine.BattleRequestCallback;
import com.trtc.uikit.livekit.livestreamcore.R;
import com.trtc.uikit.livekit.livestreamcore.common.utils.Logger;
import com.trtc.uikit.livekit.livestreamcore.manager.api.ILiveStream;
import com.trtc.uikit.livekit.livestreamcore.state.BattleState;
import com.trtc.uikit.livekit.livestreamcore.state.LiveStreamState;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class BattleManager extends BaseManager {

    private final BattleState    mBattleState;
    private       BattleObserver mBattleObserver;

    public BattleManager(LiveStreamState state, ILiveStream service) {
        super(state, service);
        mBattleState = state.battleState;
    }

    public void setBattleObserver(BattleObserver observer) {
        mBattleObserver = observer;
    }

    public BattleUser getBattleUser(String userId) {
        for (BattleUser battleUser : mBattleState.mBattledUsers.get()) {
            if (TextUtils.equals(userId, battleUser.userId)) {
                return battleUser;
            }
        }
        return null;
    }

    public void requestBattle(BattleConfig config, List<String> roomIdList, int timeout,
                              BattleRequestCallback callback) {
        mVideoLiveService.requestBattle(config, roomIdList, timeout, new TUILiveBattleManager.BattleRequestCallback() {
            @Override
            public void onSuccess(TUILiveBattleManager.BattleInfo battleInfo,
                                  Map<String, TUILiveBattleManager.BattleCode> map) {
                mBattleState.battleId = battleInfo.battleId;
                List<String> sentBattleRequestList = mBattleState.sentBattleRequestList.get();
                for (Map.Entry<String, TUILiveBattleManager.BattleCode> entry : map.entrySet()) {
                    String key = entry.getKey();
                    TUILiveBattleManager.BattleCode code = entry.getValue();
                    if (code == TUILiveBattleManager.BattleCode.SUCCESS) {
                        for (ConnectionUser user : mVideoLiveState.coHostState.connectedUserList.get()) {
                            if (TextUtils.equals(user.userId, key)) {
                                sentBattleRequestList.add(user.userId);
                                break;
                            }
                        }
                    } else {
                        BattleErrorHandler.onError(code);
                    }
                }
                if (callback != null) {
                    callback.onSuccess(battleInfo.battleId, sentBattleRequestList);
                }
            }

            @Override
            public void onError(TUICommonDefine.Error error, String s) {
                mBattleState.battleId = "";
                mBattleState.sentBattleRequestList.get().clear();
                if (callback != null) {
                    callback.onError(TUICommonDefine.Error.FAILED, s);
                }
            }
        });
    }

    public void cancelRequest(String battleId, List<String> userIdList, TUIRoomDefine.ActionCallback callback) {
        List<String> sendUserList = mBattleState.sentBattleRequestList.get();
        List<String> validUserIdList = new ArrayList<>();
        for (String userId : userIdList) {
            for (String sendUserId : sendUserList) {
                if (TextUtils.equals(sendUserId, userId)) {
                    validUserIdList.add(userId);
                    break;
                }
            }
        }
        mVideoLiveService.cancelBattleRequest(battleId, validUserIdList, callback);
    }

    public void acceptBattle(String battleId, TUIRoomDefine.ActionCallback callback) {
        mVideoLiveService.acceptBattle(battleId, callback);
    }

    public void rejectBattle(String battleId, TUIRoomDefine.ActionCallback callback) {
        mVideoLiveService.rejectBattle(battleId, callback);
    }

    public void exitBattle(String battleId, TUIRoomDefine.ActionCallback callback) {
        mVideoLiveService.exitBattle(battleId, callback);
    }

    public void onBattleStarted(BattleInfo battleInfo) {
        List<BattleUser> battleUserList = mBattleState.mBattledUsers.get();
        battleUserList.add(battleInfo.inviter);
        battleUserList.addAll(battleInfo.inviteeList);
        mBattleState.mBattledUsers.set(battleUserList);
        if (mBattleObserver != null) {
            mBattleObserver.onBattleStarted(battleInfo);
        }
    }

    public void onBattleEnded(BattleInfo battleInfo, BattleStoppedReason reason) {
        mBattleState.mBattledUsers.clear();
        if (mBattleObserver != null) {
            mBattleObserver.onBattleEnded(battleInfo);
        }
    }

    public void onUserJoinBattle(String battleId, BattleUser battleUser) {
        if (mBattleObserver != null) {
            mBattleObserver.onUserJoinBattle(battleId, battleUser);
        }
    }

    public void onUserExitBattle(String battleId, BattleUser battleUser) {
        if (mBattleObserver != null) {
            mBattleObserver.onUserExitBattle(battleId, battleUser);
        }
    }

    public void onBattleScoreChanged(String battleId, List<BattleUser> battleUserList) {
        if (mBattleObserver != null) {
            mBattleObserver.onBattleScoreChanged(battleId, battleUserList);
        }
    }

    public void onBattleRequestReceived(BattleInfo battleInfo, BattleUser inviter, BattleUser invitee) {
        if (mBattleObserver != null) {
            mBattleObserver.onBattleRequestReceived(battleInfo.battleId, inviter, invitee);
        }
    }

    public void onBattleRequestCancelled(BattleInfo battleInfo, BattleUser inviter, BattleUser invitee) {
        if (mBattleObserver != null) {
            mBattleObserver.onBattleRequestCancelled(battleInfo.battleId, inviter, invitee);
        }
    }

    public void onBattleRequestTimeout(BattleInfo battleInfo, BattleUser inviter, BattleUser invitee) {
        if (mBattleObserver != null) {
            mBattleObserver.onBattleRequestTimeout(battleInfo.battleId, inviter, invitee);
        }
    }

    public void onBattleRequestAccept(BattleInfo battleInfo, BattleUser inviter, BattleUser invitee) {
        if (mBattleObserver != null) {
            mBattleObserver.onBattleRequestAccept(battleInfo.battleId, inviter, invitee);
        }
    }

    public void onBattleRequestReject(BattleInfo battleInfo, BattleUser inviter, BattleUser invitee) {
        if (mBattleObserver != null) {
            mBattleObserver.onBattleRequestReject(battleInfo.battleId, inviter, invitee);
        }
    }

    @Override
    public void destroy() {

    }

    public interface BattleObserver {
        void onBattleStarted(BattleInfo battleInfo);

        void onBattleEnded(BattleInfo battleInfo);

        void onUserJoinBattle(String battleId, BattleUser battleUser);

        void onUserExitBattle(String battleId, BattleUser battleUser);

        void onBattleScoreChanged(String battleId, List<BattleUser> battleUserList);

        void onBattleRequestReceived(String battleId, BattleUser inviter, BattleUser invitee);

        void onBattleRequestCancelled(String battleId, BattleUser inviter, BattleUser invitee);

        void onBattleRequestTimeout(String battleId, BattleUser inviter, BattleUser invitee);

        void onBattleRequestAccept(String battleId, BattleUser inviter, BattleUser invitee);

        void onBattleRequestReject(String battleId, BattleUser inviter, BattleUser invitee);
    }

    private static class BattleErrorHandler {

        public static void onError(BattleCode code) {
            if (code == BattleCode.SUCCESS || code == null) {
                return;
            }
            String message = convertCodeToString(code);
            Logger.info("BattleErrorHandler :[code:" + code + ",message:" + message + "]");
            showToast(message);
        }

        private static String convertCodeToString(BattleCode code) {
            Context context = ContextProvider.getApplicationContext();
            switch (code) {
                case BATTLING:
                case BATTLING_OTHER_ROOM:
                    return context.getString(R.string.livestreamcore_battle_error_conflict);
                default:
                    return context.getString(R.string.livestreamcore_battle_error_other);
            }
        }

        private static void showToast(String tips) {
            Context context = ContextProvider.getApplicationContext();
            View view = LayoutInflater.from(context).inflate(R.layout.livestreamcore_connection_toast, null, false);

            TextView text = view.findViewById(R.id.tv_toast_text);
            text.setText(tips);
            ImageView image = view.findViewById(R.id.iv_toast_image);
            image.setImageResource(R.drawable.livestreamcore_connection_toast_icon);

            Toast toast = new Toast(view.getContext());
            toast.setDuration(Toast.LENGTH_SHORT);
            toast.setView(view);
            toast.show();
        }
    }
}
